src/python/pants/engine/internals/scheduler.py

```diff
@@ -522,6 +522,12 @@ def product_request(
         :param timeout: See self.execution_request.
         :returns: A list of the requested products, with length match len(subjects).
         """
+        # NB: It's possible to have synchronous requests still running after the console has
+        # shutdown, and those will crash if they attempt to write to the dynamic UI.
+        #
+        # The only time we want to use the dynamic UI is possibly when running goal rules with the
+        # asynchronous Rules API. Here, it's safest to always ensure we are logging.
+        self.teardown_dynamic_ui()
```

Comment:
I think that this has the potential to race with other requests in the `Session`... although we know that all of the calls to `run_goal_rule` will have completed before the `TC` plugin calls `get_expanded_specs`, another implementation of an SWH might call `get_expanded_specs` at the beginning of the run, which would break the UI rendering in the foreground.

I would suggest moving the [should_render_ui](https://github.com/pantsbuild/pants/blob/172b6d00ce0f30f848bc317c0b5307cbdd5029e7/src/rust/engine/src/session.rs#L101) flag to `ExecutionRequest`, and then making the `Session's` [display](https://github.com/pantsbuild/pants/blob/172b6d00ce0f30f848bc317c0b5307cbdd5029e7/src/rust/engine/src/session.rs#L67-L68) field optional:
```
struct Session {
  ..
  display: Mutex<Option<SessionDisplay>>,
  .. 
```
(or as an alternative to an option, adding a "SessionDisplay::NotRunning" variant).

The methods of `Session` that interact with the UI would then noop if there was no display running, and `maybe_display_initialize` would take the "should_render_ui" flag rather than the constructor.

Reply:
> another implementation of an SWH might call get_expanded_specs at the beginning of the run, which would break the UI rendering in the foreground.

Hm, this would suffer from the same race if we move this onto `ExecutionRequest`, right? We would be still calling the shutdown code in https://github.com/pantsbuild/pants/blob/172b6d00ce0f30f848bc317c0b5307cbdd5029e7/src/python/pants/engine/streaming_workunit_handler.py#L80-L90, only hardcoding to do teardown for only those ERs rather than every single one. But regardless of whether the SWH is finished=True, we would be saying that ER is not safe to use the dynamic UI, right?

Or perhaps something like only set on the `ER` if `finished=True`, change the API to `context.get_expanded_specs(finished: bool)`?

Reply:
> Hm, this would suffer from the same race if we move this onto ExecutionRequest, right? We would be still calling the shutdown code in

It comes down to how the `Session` methods that interact with a UI behave: https://github.com/pantsbuild/pants/blob/172b6d00ce0f30f848bc317c0b5307cbdd5029e7/src/rust/engine/src/session.rs#L209-L267 ... it might be the case that only the `ConsoleUI` needs to be under the `Mutex`, and the `SessionDisplay` would move to just being a value on the stack in `Scheduler::execute`.

So it would be
```
struct Session {
  ..
  display: Mutex<Option<ConsoleUI>>,
  .. 
}
```
... and the `Session::maybe_display_initialize` method would _return_ a `SessionDisplay` enum to drive what happens for that run of `Scheduler::execute`.